### PR TITLE
Add a metadata service to store file metadata

### PR DESCRIPTION
**Pull request by CarlSchwan**
- [x] requires https://github.com/nextcloud/3rdparty/pull/1023
   -  [x] use new release instead of patch

### CODE REVIEW

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php [hunk 1]
```diff
@@ -448,6 +473,29 @@ public function handleGetProperties(PropFind $propFind, \Sabre\DAV\INode $node)
...
+					if (str_starts_with($child->getFileInfo()->getMimeType(), 'image')) {
+						/** @var File $child */
+						$fileIds[] = $child->getFileInfo()->getId();
+					}
+				}
+				/** @var IMetaDataManager $metadataManager */
+				$metadataManager = \OC::$server->get(IMetadataManager::class);
+				$preloadedMetadata = $metadataManager->fetchMetadataFor('size', $fileIds);
+				foreach ($node->getChildren() as $child) {
+					if (str_starts_with($child->getFileInfo()->getMimeType(), 'image')) {
```
Comment (PVince81) [inline apps/dav/lib/Connector/Sabre/FilesPlugin.php:492]: should start with "image/" not just "image" ?
or you can use getMimePart() for just the first part I think

File: core/Application.php [hunk 2]
```diff
@@ -301,5 +305,10 @@ function (GenericEvent $event) use ($container) {
 		$eventDispatcher->addServiceListener(BeforeUserDeletedEvent::class, UserDeletedFilesCleanupListener::class);
 		$eventDispatcher->addServiceListener(UserDeletedEvent::class, UserDeletedFilesCleanupListener::class);
 		$eventDispatcher->addServiceListener(UserDeletedEvent::class, UserDeletedWebAuthnCleanupListener::class);
+
+		// Metadata
+		$eventDispatcher = \OC::$server->get(IEventDispatcher::class);
```
Comment (PVince81) [inline core/Application.php:310]: I really wonder if we should add a switch.

Some users might not even use the web UI and might not want metadata to be extracted at all.
In config.php we already have a switch to disable previews, so this would fit also.

File: core/Migrations/Version240000Date20220404230027.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		if (!$schema->hasTable('file_metadata')) {
+			$table = $schema->createTable('file_metadata');
+			$table->addColumn('id', 'integer', [
+				'notnull' => true,
+			]);
+			$table->addColumn('group_name', 'string', [
+				'notnull' => true,
+				'length' => 50,
+			]);
+			$table->addColumn('data', 'string', [
```
Comment (PVince81) [inline core/Migrations/Version240000Date20220404230027.php:36]: in theory there's a column type called "json", but I'm not sure if it works on all DBMS

File: core/Migrations/Version240000Date20220404230027.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+				'notnull' => true,
+				'length' => 2000,
```
Comment (PVince81) [inline core/Migrations/Version240000Date20220404230027.php:38]: should we make that 4000 right away, considering future metadata types ?
increasing the column size later might be costy if the table contains lots of entries

File: lib/private/Metadata/Provider/ExifProvider.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+				'width' => $data['COMPUTED']['Width'],
+				'height' => $data['COMPUTED']['Height'],
+			]);
+		}
+
+		if (array_key_exists('GPS', $data)) {
+			$gps->setMetadata([
+				'longitude' => $this->getGps($data['GPS']['GPSLongitude'],
+					$data['GPS']['GPSLongitudeRef']),
+				'latitude' => $this->getGps($data['GPS']['GPSLatitude'],
```
Comment (PVince81) [inline lib/private/Metadata/Provider/ExifProvider.php:84]: no altitude ? :airplane:  :smile:

File: lib/private/Server.php [hunk 6]
```diff
@@ -1427,6 +1429,8 @@ public function __construct($webRoot, \OC\Config $config) {
 
 		$this->registerAlias(IBroker::class, Broker::class);
 
+		$this->registerAlias(IMetadataManager::class, MetaDataManager::class);
```
Comment (PVince81) [inline lib/private/Server.php:1432]: this is actually "file metadata manager", not sure if we can reuse this

probably ok to keep the name

File: lib/private/Metadata/MetaDataManager.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace OC\Metadata;
+
+use OC\Metadata\Provider\ExifProvider;
+use OCP\Files\File;
+use OCP\Files\Storage\IStorage;
+use OCP\IConfig;
+use Psr\Log\LoggerInterface;
+
+class MetaDataManager implements IMetadataManager {
```
Comment (PVince81) [inline lib/private/Metadata/MetaDataManager.php:11]: inconsistent casing. I suggest we stick with lower case "d" for data

File: lib/private/Metadata/MetaDataManager.php [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	public function fetchMetadataFor(string $group, array $fileIds): array {
+		return $this->fileMetadataMapper->findForGroupForFiles($fileIds, $group);
+	}
+
+	public function getCapabilities(): array {
+		$capabilities = [];
+		foreach ($this->providers as $supportedMimetype => $provider) {
+			$capabilities[$supportedMimetype] = $provider::groupsProvided();
```
Comment (PVince81) [inline lib/private/Metadata/MetaDataManager.php:81]: if we add a switch for the extraction, we'll also need to disable the capability so clients can fall back

File: lib/private/Metadata/IMetadataManager.php [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+use OCP\Files\File;
+
+interface IMetadataManager {
+	/**
+	 * @param class-string<IMetadataProvider> $className
+	 */
+	public function registerProvider(string $className): void;
+
+	public function generateMetadata(File $node, bool $checkExisting): void;
```
Comment (PVince81) [inline lib/private/Metadata/IMetadataManager.php:13]: add some PHP docs here

File: lib/private/Metadata/FileMetadataMapper.php [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+	/**
+	 * @return array<int, FileMetadata>
+	 * @throws Exception
+	 */
+	public function findForGroupForFiles(array $fileIds, string $groupName): array {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where($qb->expr()->in('id', $qb->createNamedParameter($fileIds,IQueryBuilder::PARAM_INT_ARRAY)))
```
Comment (PVince81) [inline lib/private/Metadata/FileMetadataMapper.php:53]: chunk this to avoid reaching 1000 limit or Oracle

File: lib/private/Metadata/FileEventListener.php [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+		if ($node->getMimetype() === 'httpd/unix-directory') {
+			return false;
+		}
+		if ($node->getSize(false) <= 0) {
+			return false;
+		}
+
+		$path = $node->getPath();
+		// TODO make this more dynamic, we have the same issue in other places
+		return !str_starts_with($path, 'appdata_') && !str_starts_with($path, 'files_versions/') && !str_starts_with($path, 'files_trashbin/');
```
Comment (PVince81) [inline lib/private/Metadata/FileEventListener.php:26]: I think we have constants also for the virtual folders (future feature), in which case I think metadata would be already extracted through other means. I'm not sure if the scanner even goes there. @icewind1991 might know

File: lib/private/Metadata/Provider/ExifProvider.php [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+			return $parts[0];
+		}
+
+		return floatval($parts[0]) / floatval($parts[1]);
+	}
+
+
+	public function execute(File $file): array {
+		$fileDescriptor = $file->fopen('rb');
+		$data = exif_read_data($fileDescriptor, 'ANY_TAG', true);
```
Comment (PVince81) [inline lib/private/Metadata/Provider/ExifProvider.php:51]: add error handling, we should not crash in case of error

now, should we store `{}` still so we don't attempt again repeatedly ?
repeatedly might not happen anyway since it only happens at file creation :thinking:

Comment (PVince81) [reply lib/private/Metadata/Provider/ExifProvider.php:51]: ah, you got it below. maybe move the `if (!$data)` higher

File: lib/private/Files/Cache/Scanner.php [hunk 13]
```diff
@@ -57,62 +76,41 @@
...
-	 * @var bool $useTransactions whether to use transactions
-	 */
-	protected $useTransactions = true;
-
-	/**
-	 * @var \OCP\Lock\ILockingProvider
-	 */
-	protected $lockingProvider;
+	protected ?IStorage $storage = null;
+	protected ?string $storageId = null;
```
Comment (PVince81) [inline lib/private/Files/Cache/Scanner.php:80]: I'm a bit afraid, remember that other PR where we added a lot of type hinting and stuff started to break ?
and this class here is very central to Nextcloud, should we limit type hinting to new additions ?

File: lib/private/Files/Cache/Scanner.php [hunk 14]
```diff
@@ -140,149 +138,136 @@ public function scanFile($file, $reuseExisting = 0, $parentId = -1, $cacheData =
...
-		return null;
+		if (!isset($data['encrypted'])) {
+			$data['encrypted'] = false;
+		}
+		return $data;
 	}
 
 	protected function removeFromCache($path) {
+		$node = $this->getNodeForPath($path);
+		echo 'ree';
```
Comment (PVince81) [inline lib/private/Files/Cache/Scanner.php:257]: stray debugging code

File: lib/private/Files/Cache/Scanner.php [hunk 15]
```diff
@@ -140,149 +138,136 @@ public function scanFile($file, $reuseExisting = 0, $parentId = -1, $cacheData =
...
+		}
 
-			//acquire a lock
-			if ($lock) {
-				if ($this->storage->instanceOfStorage('\OCP\Files\Storage\ILockingStorage')) {
-					$this->storage->acquireLock($file, ILockingProvider::LOCK_SHARED, $this->lockingProvider);
-				}
+		// Acquire a lock
+		if ($lock) {
+			if ($this->storage->instanceOfStorage(ILockingStorage::class)) {
```
Comment (PVince81) [inline lib/private/Files/Cache/Scanner.php:148]: this is ok, please note that for some classes we have to keep string, like ISharedStorage.
so if you ever encounter such strings, always double check if they are part of an app that can be disabled

**Comment by juliushaertl** [comment]
Just curious, is searching for files by metadata something in scope? I could imagine this a useful scenario for future enhancements of the photos app to search by location for example, as with JSON-only metadata db field that would not scale.

**Comment by CarlSchwan** [comment]
> Just curious, is searching for files by metadata something in scope? I could imagine this a useful scenario for future enhancements of the photos app to search by location for example, as with JSON-only metadata db field that would not scale.

Unfortunately, this seems to be a bit tricky. If we want to search photos by location in an efficient way we need more than just the simple GPS coordinate but we need to map this to a geographic area. This is a bit tricky to implement but we can inspire outself from  something similar I have in Koko: https://invent.kde.org/graphics/koko/-/blob/master/src/reversegeocoder.cpp#L30 (this requires to be done in a background job as loading the data is quite expensive already in pure C/C++ with an optimized kdtree data structure)

This geographic area should probably be stored in a separate table e.g. https://invent.kde.org/graphics/koko/-/blob/master/src/imagestorage.cpp#L132

But the biggest issue is that while this is fairly simple to implement in the one user use-case, for a Nextcloud we need to retrieve the list of locations or the list of photos in one location on a per user basis and this is really tricky unless we limit ourself only to files stored in the user storage (so no group folders, no sharing, no circles ,...). @PVince81 @AndyScherzinger  Would this is be ok?

In that case, we could adopt the following DB structure for location information.

```SQL
CREATE TABLE oc_file_metadata (
   id INTEGER,
   group TEXT,
   data TEXT, # nullable
   foreign INTEGER, # nullable
   UNIQUE(id, group)
)

CREATE TABLE oc_location (
    id INTEGER PRIMARY KEY,
    country TEXT,
    state TEXT,
    city TEXT,
    UNIQUE(country, state, city)
)
```

The PHP API needs to be expanded to describe if the data is directly retrievable in the JSON data column, or to be joined with another table (e.g for locations with the oc_location table). To be able only to retrieve the locations for one user, it should be possible to join with oc_storage and oc_filecache to get only the ids of files from the user. This might still be expensive when a user has a log of photos but I don't really see a better way to do it, unfortunately :(

**Comment by AndyScherzinger** [comment]
I think the location data needs a more detailed discussion, so from my perspective we can skip that for now. The only thing we "need" for 24 is height/width for the Clients to implement a different layout for their media views. (kinda like a gallery).

So if there if leaving location out doesn't create a future blocker just skip it for now. Because this needs a lot more thinking from my perspective because with with/height it is just extraction+exposure, for anything else it is likely more than that, like "places" (like you also outlined), "people" only other simple info I can come up with would be certain exif data (like exposure, aperture, cam-model, lense) but since we also don't make use of that for now, this can also be skipped.

So from my point of view for v24 - keep it simple by limiting it to width/height only. What do you think @PVince81 ?

**Comment by PVince81** [comment]
yes, keep it simple for now.

if we need to cache/compute more data for GPS search later, it will be a separate task anyway

**Comment by PVince81** [comment]
had a quick chat with @CarlSchwan about the location data:
1) I don't think it's a good idea to link oc_location to oc_file_metadata because the latter table also contains non-GPS data, so it feels a bit weird to link them there

2) an alternative would be to connect oc_locations with oc_filecache. Since it's a 1-N relationship, we'd need a new table oc_location_mapping or similar to connect: oc_filecache->oc_location_mapping->oc_locations. This way it becomes possible to find all file ids for a specific location

We both agreed that 2) is a better approach and will remove the need to add anything else to oc_file_metadata, thus deblocking this PR.

File: lib/private/Metadata/FileMetadataMapper.php [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+	 * @throws Exception
+	 */
+	public function findForGroupForFiles(array $fileIds, string $groupName): array {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where($qb->expr()->in('id', $qb->createParameter('fileIds')))
+			->andWhere($qb->expr()->eq('group_name', $qb->createNamedParameter($groupName, IQueryBuilder::PARAM_STR)));
+
+		foreach (array_chunk($fileIds) as $fileIdsChunk) {
```
Comment (PVince81) [inline lib/private/Metadata/FileMetadataMapper.php:56]: too few arguments?

Comment (CarlSchwan) [reply lib/private/Metadata/FileMetadataMapper.php:56]: Yeah 😅 i have a fix locally, i push it with the other fixes for your comments

File: lib/private/Metadata/Provider/ExifProvider.php [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+use OCP\Files\File;
+use OCP\IConfig;
+use Psr\Log\LoggerInterface;
+
+class ExifProvider implements IMetadataProvider {
+	public function __construct(IConfig $config, LoggerInterface $logger) {
+	}
+
+	static public function groupsProvided(): array {
+		return ['size', 'gps'];
```
Comment (PVince81) [inline lib/private/Metadata/Provider/ExifProvider.php:16]: please check the warnings

File: lib/private/Files/Cache/Scanner.php (hunk 13 above)
Comment (CarlSchwan) [reply lib/private/Files/Cache/Scanner.php:80]: `?type $name = null` is not that dangerous compared to `type $name` as it will just create an error when the type is not correct instead of an error when the variable is empty

File: core/Migrations/Version240000Date20220404230027.php (hunk 3 above)
Comment (CarlSchwan) [reply core/Migrations/Version240000Date20220404230027.php]: json is a nice idea :) doctrine takes care of the compatibility here by using TEXT when it's not available but from my research it seems to work on all the DB we support

File: lib/private/Metadata/Provider/ExifProvider.php (hunk 5 above)
Comment (CarlSchwan) [reply lib/private/Metadata/Provider/ExifProvider.php]: I commented this code for now

File: lib/private/Metadata/FileEventListener.php (hunk 11 above)
Comment (CarlSchwan) [reply lib/private/Metadata/FileEventListener.php:43]: the events are trigerred for these code paths :(

File: lib/private/Files/Cache/Scanner.php (hunk 13 above)
Comment (CarlSchwan) [reply lib/private/Files/Cache/Scanner.php]: reverted most of my changes to the scanner. let's move that to another PR later one

**Comment by PVince81** [comment]
- [ ] update config.sample.php with the new config switch @CarlSchwan

File: core/Application.php [hunk 18]
```diff
@@ -301,5 +306,14 @@ function (GenericEvent $event) use ($container) {
 		$eventDispatcher->addServiceListener(BeforeUserDeletedEvent::class, UserDeletedFilesCleanupListener::class);
 		$eventDispatcher->addServiceListener(UserDeletedEvent::class, UserDeletedFilesCleanupListener::class);
 		$eventDispatcher->addServiceListener(UserDeletedEvent::class, UserDeletedWebAuthnCleanupListener::class);
+
+		// Metadata
+		/** @var IConfig $config */
+		$config = $container->get(IConfig::class);
+		if ($config->getSystemValueBool('enable-metadata', true)) {
```
Comment (PVince81) [inline core/Application.php:313]: ah, too bad we don't have "files.xxxx", otherwise "files.enable_metadata" would have been nice
we have "sharing.xxxx" settings

then how about "enable_files_metadata" ?

in any case, use underscores like other settings in config.sample.php

File: core/Migrations/Version240000Date20220404230027.php [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+		if (!$schema->hasTable('file_metadata')) {
+			$table = $schema->createTable('file_metadata');
+			$table->addColumn('id', Types::INTEGER, [
+				'notnull' => true,
+			]);
+			$table->addColumn('group_name', Types::STRING, [
+				'notnull' => true,
+				'length' => 50,
+			]);
+			$table->addColumn('metadata', Types::JSON, [
```
Comment (PVince81) [inline core/Migrations/Version240000Date20220404230027.php:53]: curious to see if Oracle will pass :sweat_smile:

File: lib/private/Files/Cache/Scanner.php [hunk 20]
```diff
@@ -87,12 +100,19 @@ class Scanner extends BasicEmitter implements IScanner {
...
+
 	public function __construct(\OC\Files\Storage\Storage $storage) {
 		$this->storage = $storage;
 		$this->storageId = $this->storage->getId();
 		$this->cache = $storage->getCache();
 		$this->cacheActive = !\OC::$server->getConfig()->getSystemValue('filesystem_cache_readonly', false);
 		$this->lockingProvider = \OC::$server->getLockingProvider();
+		$this->eventDispatcher = \OC::$server->get(IEventDispatcher::class);
+		$this->root = \OC::$server->get(IRootFolder::class);
+		$this->view = Filesystem::getView();
```
Comment (PVince81) [inline lib/private/Files/Cache/Scanner.php:115]: I think this violates the layers.
View is supposed to be on top :-S

if we want to to bypass this we'd need to levels of hooks: internal legacy hooks that trigger with the file id, and then further resolve it

or change the events to just pass a file id and resolve the node outside ?

File: lib/private/Metadata/Provider/ExifProvider.php [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+		$minutes = count($exifCoordinate) > 1 ? $this->gps2Num($exifCoordinate[1]) : 0;
+		$seconds = count($exifCoordinate) > 2 ? $this->gps2Num($exifCoordinate[2]) : 0;
+
+		$flip = ($hemi == 'W' or $hemi == 'S') ? -1 : 1;
+
+		return (float)($flip * ($degrees + $minutes / 60 + $seconds / 3600));
+
+	}
+
+	private function gps2Num(string $coordinatePart): float {
```
Comment (PVince81) [inline lib/private/Metadata/Provider/ExifProvider.php:35]: should we move this to another PR maybe ?

I like that it illustrates the possibilities but someone might be mistaken to think they can simply uncomment

File: lib/private/Files/Cache/Scanner.php (hunk 20 above)
Comment (CarlSchwan) [reply lib/private/Files/Cache/Scanner.php]: the view is now created on demand later on

File: lib/private/Files/Cache/Scanner.php [hunk 22]
```diff
@@ -565,4 +587,29 @@ private function runBackgroundScanJob(callable $callback, $path) {
 	public function setCacheActive($active) {
 		$this->cacheActive = $active;
 	}
+
+	private function getNodeForPath(string $path): Node {
+		if (!$this->view) {
```
Comment (PVince81) [inline lib/private/Files/Cache/Scanner.php:592]: add a "FIXME: move this to higher layers" at least, to acknowledge that it's out of place

File: lib/private/Files/Cache/Scanner.php [hunk 23]
```diff
@@ -87,12 +100,19 @@ class Scanner extends BasicEmitter implements IScanner {
...
+
 	public function __construct(\OC\Files\Storage\Storage $storage) {
 		$this->storage = $storage;
 		$this->storageId = $this->storage->getId();
 		$this->cache = $storage->getCache();
 		$this->cacheActive = !\OC::$server->getConfig()->getSystemValue('filesystem_cache_readonly', false);
 		$this->lockingProvider = \OC::$server->getLockingProvider();
+		$this->eventDispatcher = \OC::$server->get(IEventDispatcher::class);
+		$this->root = \OC::$server->get(IRootFolder::class);
+		$this->view;
```
Comment (PVince81) [inline lib/private/Files/Cache/Scanner.php:115]: remove this row that does nothing ?

File: core/Migrations/Version240000Date20220404230027.php (hunk 19 above)
Comment (CarlSchwan) [reply core/Migrations/Version240000Date20220404230027.php:53]: https://docs.oracle.com/database/121/ADXDB/json.htm#ADXDB6246

it does and probably store it internally as Cbor so it's more space efficiant

**Comment by tobiasKaminsky** [comment]
@CarlSchwan can I test this somehow already?
How can I regenerate metadata for existing images?

File: lib/private/Files/Cache/Scanner.php [hunk 24]
```diff
@@ -269,11 +289,14 @@ public function scanFile($file, $reuseExisting = 0, $parentId = -1, $cacheData =
 	}
 
 	protected function removeFromCache($path) {
+		$node = $this->getNodeForPath($path);
+		$this->eventDispatcher->dispatchTyped(new BeforeNodeDeletedEvent($node));
```
Comment (icewind1991) [inline lib/private/Files/Cache/Scanner.php:293]: Why is this emitted here instead of from higher layers?

**Comment by CarlSchwan** [comment]
> @CarlSchwan can I test this somehow already? How can I regenerate metadata for existing images?

Just run the branch and you should be able to see the metadata with the <file-metadata-size> PROPFIND/SEARCH parameter.

Unfortunately, the metadata regeneration job is not implemented yet and is task for another PR (probably not 24.0.0) so you need to handle the case where the metadata are not available

File: lib/private/Files/Cache/Scanner.php (hunk 24 above)
Comment (CarlSchwan) [reply lib/private/Files/Cache/Scanner.php]: I changed that to listening to NodeRemovedFromCache instead in the FileListener

**Comment by tobiasKaminsky** [comment]
> Unfortunately, the metadata regeneration job is not implemented yet and is task for another PR (probably not 24.0.0) so you need to handle the case where the metadata are not available

So it might be that clients receive e.g. 500 images, but only 200 images have a valid metadata.
Is this something clients can now before? 
Media searches are done in batches, so it might be that first batch has valid metadata and thus client shows new view (which uses metadata), but then second batch is without valid metadata and thus the view is mixed up…

@jancborchardt ^ as we spoke in our call about it, but I have no idea how to display this hybrid view

**Comment by jancborchardt** [comment]
@CarlSchwan @PVince81 btw what is the reason for the metadata capability being optional? Customer requirement (don’t see that in the customer issue), not running on all setups, or something else?
(Just asking cause wouldn’t it be easier and more consistent to just always have it?)

**Comment by CarlSchwan** [comment]
> @CarlSchwan @PVince81 btw what is the reason for the metadata capability being optional? Customer requirement (don’t see that in the customer issue), not running on all setups, or something else? (Just asking cause wouldn’t it be easier and more consistent to just always have it?)

Also we want to support in the future the possibility for apps to provide their own metadata extractor and dynamically indicate in the capability that metadata is available for each mimetype

**Review by PVince81 — APPROVED** [review]
:+1: 

please see my one comment about the interface

File: lib/private/Metadata/IMetadataProvider.php [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+
+use OCP\Files\File;
+use OCP\IConfig;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Interface for the metadata providers. If you want an application to provide
+ * some metadata, you can use this to store them.
+ */
+interface IMetadataProvider {
```
Comment (icewind1991) [inline lib/private/Metadata/IMetadataProvider.php:13]: Is there a reason for these interfaces not to be in `OCP`?

Comment (CarlSchwan) [reply lib/private/Metadata/IMetadataProvider.php:13]: We didn't want to have a stable api for 24, but it's a goal for 25 to move that to ocp

File: lib/private/Metadata/IMetadataProvider.php [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+	public function __construct(IConfig $config, LoggerInterface $logger);
```
[the remark this answers is not in the document]
Comment (CarlSchwan) [reply lib/private/Metadata/IMetadataProvider.php:14]: right, we could probably do dependy injection, this would be a bit nicer

**Comment by tobiasKaminsky** [comment]
@CarlSchwan if there is none metadata, "[ ]" is returned, but with metadata it is:
 {"width":960,"height":1280}

Json parser gets a hiccup here, as it expects "{}" in case of empty.

**Comment by CarlSchwan** [comment]
> @CarlSchwan if there is none metadata, "[ ]" is returned, but with metadata it is: {"width":960,"height":1280}
> 
> Json parser gets a hiccup here, as it expects "{}" in case of empty.

@tobiasKaminsky this should be fixed. Strange that the json parser doesn't support [] :(

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php [hunk 27]
```diff
@@ -436,6 +440,29 @@ public function handleGetProperties(PropFind $propFind, \Sabre\DAV\INode $node)
...
+						$sizeMetadata = $node->getMetadata('size');
+					} else {
+						// This code path should not be called since we try to preload
+						// the metadata when loading the folder or the search results
+						// in one go
+						$metadataManager = \OC::$server->get(IMetadataManager::class);
+						$sizeMetadata = $metadataManager->fetchMetadataFor('size', [$node->getId()])[$node->getId()];
+
+						// TODO would be nice to display this in the profiler...
+						\OC::$server->get(LoggerInterface::class)->warning('Inefficient fetching of metadata');
```
Comment (icewind1991) [inline apps/dav/lib/Connector/Sabre/FilesPlugin.php:460]: This should be `debug` instead of `warning` to prevent adding a bunch of logs for users who can't do anything to fix this anyway

File: apps/dav/lib/Files/FileSearchBackend.php [hunk 28]
```diff
@@ -124,23 +124,52 @@ public function getPropertyDefinitionsForScope($href, $path) {
...
-			new SearchPropertyDefinition('{DAV:}resourcetype', false, true, false),
-			new SearchPropertyDefinition('{DAV:}getcontentlength', false, true, false),
-			new SearchPropertyDefinition(FilesPlugin::CHECKSUMS_PROPERTYNAME, false, true, false),
-			new SearchPropertyDefinition(FilesPlugin::PERMISSIONS_PROPERTYNAME, false, true, false),
-			new SearchPropertyDefinition(FilesPlugin::GETETAG_PROPERTYNAME, false, true, false),
-			new SearchPropertyDefinition(FilesPlugin::OWNER_DISPLAY_NAME_PROPERTYNAME, false, true, false),
-			new SearchPropertyDefinition(FilesPlugin::DATA_FINGERPRINT_PROPERTYNAME, false, true, false),
-			new SearchPropertyDefinition(FilesPlugin::HAS_PREVIEW_PROPERTYNAME, false, true, false, SearchPropertyDefinition::DATATYPE_BOOLEAN),
-			new SearchPropertyDefinition(FilesPlugin::FILEID_PROPERTYNAME, false, true, false, SearchPropertyDefinition::DATATYPE_NONNEGATIVE_INTEGER),
+			new SearchPropertyDefinition('{DAV:}resourcetype', true, false, false),
```
Comment (icewind1991) [inline apps/dav/lib/Files/FileSearchBackend.php:127]: why the changes to `searchable` and `selectable` here?

Comment (CarlSchwan) [reply apps/dav/lib/Files/FileSearchBackend.php:127]: This is due to the fact that the oder got changed when updating to searchdav3 https://github.com/nextcloud/3rdparty/compare/d80ec1fa2dad1c3ede272583e3c4f1f77f40141b...6176112be9428026897d958dc2b558d1bde4fec2#diff-5d91cf2a29381b5473f6aaa350e50cf8e8a68d63b80918eeedd4ed793cd76cb8R58

**Review by miaulalala — APPROVED** [review]

File: apps/dav/lib/Files/FileSearchBackend.php (hunk 28 above)
Comment (icewind1991) [reply apps/dav/lib/Files/FileSearchBackend.php:127]: Hmm, I think that was done on accident :see_no_evil:

Comment (CarlSchwan) [reply apps/dav/lib/Files/FileSearchBackend.php:127]: we can revert the order of the arguments as it was before but there is no release pressure on that
